### ISSUE TITLE
Training a Keras Model with Multiple Inputs/Outputs

### ISSUE THREAD

**Issue by EmiCareOfCell44** [intel-analytics/analytics-zoo#3824]
t seems that it is not possible to train a model that accepts a Table as input/ouput because neither Sample nor Criterions accepts Table, only Tensor. I have created the BERT/Google implementation MLM in a Keras model to trying how it could work in a big Spark cluster, adding the specific modules for MLM and NSP to the available BERT Layer. But, as I see I can´t distribute the dataset in a Sample[Table]. Furthermore, the SparseCategoricalCrossEntropy does not accept Table neither.

Have you tried something related to this, or I need to create the Sample[Table] and the criterion for training MLM and NSP at the same time?

**Comment by qiuxin2012** [comment]
We also have a BERT implement, see https://github.com/intel-analytics/analytics-zoo/tree/56288f0da3c34afecb32fecd5a7ef93ca78425be/pyzoo/zoo/examples/orca/learn/bigdl/attention for details.

**Comment by jason-dai** [comment]
@EmiCareOfCell44 Are you running TensorFlow Keras, or [Keras-like API](https://analytics-zoo.readthedocs.io/en/latest/doc/UseCase/keras-api.html) for BigDL? In both cases, it is recommended to use the new [Orca](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/orca.html) library, which allows you to directly fit using Spark Dataframes (or distributed Numpy/Pandas objects wrapped inside [XShards](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html#xshards-distributed-data-parallel-python-processing)).

**Comment by EmiCareOfCell44** [comment]
@jason-dai I´m using the Keras-like API(Scala). I´ve just built the graph. The input shape is 

**Shape(List(Shape(1, seqLen), Shape(1, seqLen), Shape(1, seqLen), Shape(1, 1, 1, seqLen)))** 

And the output is 

**Shape(List(Shape(1,2), Shape(1, 512, 30000))**

Actually, I would like not to use pySpark for the moment. There is no problem building the RDD directly. I´m thinking that I can concat the input to build the Smaple[Tensor[T]] and then create  a Layer that splits the input Tensor to create for feeding the Input, maybe it is posible to optimize the Masked Lang Model in this way. But, the Bert implementation uses two loss functions. I will try a ParallelCriterion that adds the resulting losses.

The model I´ve just created(including the Embedding similarities + bias trainable parameters) is:

Total params: 110,295,602
Trainable params: 110,295,602

I´ll give you my feedback.

**Comment by qiuxin2012** [comment]
https://github.com/intel-analytics/BigDL/blob/987053d05fa55a685a25ac2e0ad17470433688fb/spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala#L410 
This  ```def apply[T: ClassTag](
        featureTensors: Array[Tensor[T]],
        labelTensors: Array[Tensor[T]])(implicit ev: TensorNumeric[T]) : Sample[T] = {```
is the same with Sample[Table, Table],  I think it's what you need.
 SampleToMiniBatch will convert these sample[Array[Tensor], Array[Tensor]] to MiniBatch[Table, Table].

**Comment by EmiCareOfCell44** [comment]
Yes, it works. I will work the repo to include examples that use this topology. The parallel criterio seems to work as well. I close the issue. Thanks
